### PR TITLE
763 parse and manipulate changes files

### LINKED ISSUE #763: Parse and Manipulate Changes Files
Ticket #517 requests to describe resolved vulnerabilities in the changelog.

In order to simply this task, the current ticket requests to add a dedicated class for that:
* parse from raw text (input)
* render as raw text (output)
* check if a specific section is contained
* add or update a specific section
* remove or replace a specific section

Also in the past (e.g. ticket #505) an implementation has been added to include a report of dependency changes in the changelog which could benefit from a central changelog manipulation class, too.

### PR DESCRIPTION
**Pull request by ckunki**
Closes #763 

# Checklist

Note: If any of the items in the checklist are not relevant to your PR, just check the box.

## For any Pull Request

Is the following correct:
* [x] the title of the Pull Request?
* [x] the title of the corresponding issue?
* [x] there are no other open [Pull Requests](../../../../pulls) for the same update/change? 
* [x] that the issue which this Pull Request fixes ("Fixes...") is mentioned?

## When Changes Were Made

Did you:
* [x] update the changelog?
* [ ] update the cookiecutter-template?
* [x] update the implementation?
* [x] check coverage and add tests: unit tests and, if relevant, integration tests?
* [ ] update the User Guide & other documentation?
* [ ] resolve any failing CI criteria (incl. Sonar quality gate)?

## When Preparing a Release

Have you:
* [ ] thought about version number (major, minor, patch)?
* [ ] checked Exasol packages for updates and resolved open vulnerabilities, if easily possible?

### CODE REVIEW

File: exasol/toolbox/util/release/markdown.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
+"""
+Class Markdown represents a file in markdown syntax with some additional
```
Comment (ArBridgeman) [inline exasol/toolbox/util/release/markdown.py]: (nit)
```suggestion
Class Markdown represents a file in Markdown syntax with some additional
```

File: exasol/toolbox/util/release/markdown.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+constraints:
+
+* The file must start with a title in the first line.
+* Each subsequent title must be of a higher level, ie. start with more "#"
```
Comment (ArBridgeman) [inline exasol/toolbox/util/release/markdown.py]: (nit)
```suggestion
* Each subsequent title must be of a higher level, i.e. start with more "#"
```

File: exasol/toolbox/util/release/markdown.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+        """
+        Insert the specified section as child at the specified position.
+        """
+
+        self.children.insert(pos, self._check(child))
+        return self
+
+    def replace_or_append_child(self, child: Markdown) -> Markdown:
+        """
+        If there is a child with the same title then replace this child
```
Comment (ArBridgeman) [inline exasol/toolbox/util/release/markdown.py]: (nit)
```suggestion
        If there is a child with the same title, then replace this child;
```

File: exasol/toolbox/util/release/markdown.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+        otherwise append the specified child.
```
Comment (ArBridgeman) [inline exasol/toolbox/util/release/markdown.py]: (nit)
```suggestion
        otherwise, append the specified child.
```

File: test/unit/util/release/markdown_test.py [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+@pytest.mark.parametrize("scenario", ALL_SCENARIOS)
+def test_add_illegal_child(scenario: Scenario):
+    testee = scenario.create_testee()
+    with pytest.raises(IllegalChild):
+        testee.add_child(ILLEGAL_CHILD)
+
+
+def test_nested():
+    testee = NESTED.create_testee()
+    assert testee.child("## Child A").child("### Grand Child") is not None
```
Comment (ArBridgeman) [inline test/unit/util/release/markdown_test.py]: (nit)

I've been seeing this pattern a bit more. Can you tell me more about the benefit of testing `is not None`?

I'd test something like this, create a similar object, or verify it was a `Markdown` result
```suggestion
    assert testee.child("## Child A").child("### Grand Child").title == "### Grand Child"
```

File: exasol/toolbox/util/release/markdown.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+        line = stream.readline()
+        if not is_title(line):
+            raise ParseError(
+                f'First line of markdown file must be a title, but is "{line}"'
+            )
+
+        section, line = cls._parse(stream, line)
+        if not line:
+            return section
+        raise ParseError(
```
Comment (ArBridgeman) [inline exasol/toolbox/util/release/markdown.py:175]: (question)

Again, I'm just curious. I'm used to custom exceptions having the messages included in the class definition. Then, you'd pass line & section as input values. What do see as the benefit of having the whole message on the line and the class not having one?

File: exasol/toolbox/util/release/markdown.py [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+            intro += line
+            line = stream.readline()
+        if is_list_item(line):
+            while line and not is_title(line):
+                items += line
+                line = stream.readline()
+        while is_title(line) and level(title) < level(line):
+            child, line = Markdown._parse(stream, title=line)
+            children.append(child)
+        return Markdown(title, intro.strip("\n"), items.strip("\n"), children), line
```
Comment (ArBridgeman) [inline exasol/toolbox/util/release/markdown.py]: (nit)
```suggestion
        return cls(title, intro.strip("\n"), items.strip("\n"), children), line
```

File: exasol/toolbox/util/release/markdown.py [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+
+
+class IllegalChild(Exception):
+    """
+    When adding a child to a parent with higher level title.
+    """
+
+
+def is_title(line: str) -> bool:
+    return bool(line) and line.startswith("#")
```
Comment (ArBridgeman) [inline exasol/toolbox/util/release/markdown.py]: (question)
I think if line is a string that this only happens when we have an empty string ("").
Do you know of other cases it would be `False`?

File: test/unit/util/release/markdown_test.py [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+    # title
+    intro
+    * item one
+    * item two
+    ## Child
+    cintro
+    - item c1
+    - item c2
+    """,
+    expected_output="""
```
Comment (ArBridgeman) [inline test/unit/util/release/markdown_test.py:53]: ❤️

File: exasol/toolbox/util/release/markdown.py [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+
+
+def is_list_item(line: str) -> bool:
+    return bool(line) and (line.startswith("*") or line.startswith("-"))
```
Comment (ArBridgeman) [inline exasol/toolbox/util/release/markdown.py]: Might be out of scope as all this is already so great ❤️ ,
would we want to treat numerical bullet points similarly?

File: exasol/toolbox/util/release/markdown.py [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+        if not is_title(line):
+            raise ParseError(
+                f'First line of markdown file must be a title, but is "{line}"'
+            )
+
+        section, line = cls._parse(stream, line)
+        if not line:
+            return section
+        raise ParseError(
+            f'Found additional line "{line}" after top-level section "{section.title}".'
```
Comment (ArBridgeman) [inline exasol/toolbox/util/release/markdown.py:176]: (nit)
For this ParseError exception, can you add an unit test?

File: exasol/toolbox/util/release/markdown.py (hunk 8 above)
Comment (ckunki) [reply exasol/toolbox/util/release/markdown.py]: I agree:
* `bool(line)` is only `False`, if `line == ""`. 
* There could be the case `line is None`, but this violates the type hint, requiring `line` to be a `str`.

I used the `and` condition to ensure `line` is a _title_, which is only the case if line is not empty and starts with a hashmark `#`. 

Would you prefer this variant? 
```python 
    return line != "" and line.startswith("#")
```

File: exasol/toolbox/util/release/markdown.py (hunk 10 above)
Comment (ckunki) [reply exasol/toolbox/util/release/markdown.py]: I can add this, if you like.

File: exasol/toolbox/util/release/markdown.py (hunk 6 above)
Comment (ckunki) [reply exasol/toolbox/util/release/markdown.py:175]: If I included the message in the class definition, then I'd need to define two separate classes here.
As a lazy and parsimonious guy I prefer reusing the same class two times 🙂 

What I like, is that the message is exactly at the position raising the exception.
This concept is also enforced by error-handling-python, which is not used here. 
Maybe I should?

File: exasol/toolbox/util/release/markdown.py (hunk 11 above)
Comment (ckunki) [reply exasol/toolbox/util/release/markdown.py:176]: here it is:

https://github.com/exasol/python-toolbox/blob/refactoring/763-Parse_and_Manipulate_Changes_Files/test/unit/util/release/markdown_test.py#L143-L154

File: exasol/toolbox/util/release/markdown.py (hunk 10 above)
Comment (ckunki) [reply exasol/toolbox/util/release/markdown.py]: Added

File: exasol/toolbox/util/release/markdown.py (hunk 8 above)
Comment (ArBridgeman) [reply exasol/toolbox/util/release/markdown.py]: Yes, I'd slightly prefer it. I think it's a bit cleaner.

File: exasol/toolbox/util/release/markdown.py (hunk 11 above)
Comment (ArBridgeman) [reply exasol/toolbox/util/release/markdown.py:176]: Thanks, sorry for missing it the first time around.

File: exasol/toolbox/util/release/markdown.py (hunk 6 above)
Comment (ArBridgeman) [reply exasol/toolbox/util/release/markdown.py:175]: No, it's good :)

**Review by ArBridgeman — DISMISSED** [review]

File: exasol/toolbox/util/release/markdown.py (hunk 8 above)
Comment (ckunki) [reply exasol/toolbox/util/release/markdown.py]: OK, I updated the implementation.

**Review by ArBridgeman — DISMISSED** [review]

**Review by ArBridgeman — APPROVED** [review]
